src/main/kotlin/com/londogard/textgen/languagemodels/LanguageModel.kt

```diff
@@ -0,0 +1,101 @@
+package com.londogard.textgen.languagemodels
+
+import com.londogard.textgen.penalties.Penalty
+import com.londogard.textgen.tokenizers.SimpleExtensibleTokenizer
+import com.londogard.textgen.tokenizers.Tokenizer
+import com.londogard.textgen.utils.SerializerUtil
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.Serializable
+
+typealias InternalLanguageModel = Map<List<Int>, List<Pair<Int, Double>>>
+typealias InternalVocabulary = Map<Int, String>
+typealias InternalReverseVocabulary = Map<String, Int>
+
+@Serializable
+class LanguageModel(
+    val tokenizer: Tokenizer,
+    val dictionary: InternalVocabulary,
+    val internalLanguageModel: InternalLanguageModel,
+    val n: Int
+) {
+    fun getReverseDictionary(): InternalReverseVocabulary = dictionary.entries.associateBy({ it.value }) { it.key }
+
+    fun getLanguageModel(): InternalLanguageModel = internalLanguageModel
+
+    @ExperimentalSerializationApi
+    fun dump(absolutePath: String): Unit = SerializerUtil.encodeLanguageModel(absolutePath, this)
+
+    companion object {
+        @ExperimentalSerializationApi
+        fun dump(languageModel: LanguageModel, absolutePath: String): Unit =
+            SerializerUtil.encodeLanguageModel(absolutePath, languageModel)
+
+        @ExperimentalSerializationApi
+        fun load(absolutePath: String): LanguageModel = SerializerUtil.decodeLanguageModelByAbsPath(absolutePath)
+
+        fun fit(
+            documents: List<String>,
+            n: Int,
+            tokenizer: Tokenizer = SimpleExtensibleTokenizer(whitespaceRegex = " "),
+            keepMinFreq: Int = 0, // 0 = at least one occurrence, 1 = at least two occurrences, etc... Applied to all ngrams
+            preprocessing: List<(String) -> String> = emptyList(),   // Example: listOf(TextNormalizer::normalize, String::toLowerCase)
+        ): LanguageModel {
+            val preprocessedDocs = documents
+                .map { subDoc -> preprocessing.fold(subDoc) { acc, preprocessor -> preprocessor(acc) } }
+            val tokens = preprocessedDocs.map(tokenizer::split)
+            val reverseDict = tokens.flatten().toHashSet().let { uniqueTokens ->
+                HashMap<String, Int>(uniqueTokens.size).apply { putAll(uniqueTokens.zip(uniqueTokens.indices)) }
+            }
+            val tokensList = tokens.map { t -> t.mapNotNull(reverseDict::get) }
+
+            var previousMatch = true
+            val ngramMap = (1..n)
+                .fold(emptyMap<List<Int>, Int>()) { acc, ngram ->
+                    when {
+                        ngram <= 2 || previousMatch ->
```

Comment:
`previousMatch` allows us to reduce time _a lot_ in some edge cases (shakespeare is such, cardsagainst is not).
We completely stop trying whenever we didn't match anything for previous ngram.
Without minFreq this doesn't help (as you match everything).